src/statmodels.jl

```diff
@@ -197,7 +197,13 @@ Coefficient of determination (R-squared).
 For a linear model, the R² is defined as ``ESS/TSS``, with ``ESS`` the explained sum of squares
 and ``TSS`` the total sum of squares.
 """
-r2(obj::StatisticalModel) = mss(obj) / deviance(obj)
+function r2(obj::StatisticalModel)
+    Base.depwarn("The default r² method for linear models is deprecated. " *
+                 "Packages should define their own methods.", :r2)
+
+    μ = meanresponse(obj)
+    mss(obj) / sum(x -> abs2(x - μ), response(obj))
```

Comment:
Good catch. That also applies to the Efron R². For the method above, since it's deprecated, we'd better keep using `deviance` for now. For the Efron R², I guess we need to multiply observations by their weights, but will it work for all kinds of models?

Reply:
Precisely: packages can handle weights through `mss(obj)` and `loglikelihood(obj)`. There's no need to provide for weights here.

Reply:
Not sure for `ProbabilityWeights`, but should work fine for the other weights supported if not mistaken. The tests should cover the weight cases for sure.

Reply:
Efron's R² does need special treatment, since it uses the sum of residuals.

Reply:
Shouldn't that be
```
mss(obj) / sum(w * abs2(x - μ) for (w, x) in zip(weights(obj), response(obj)))
```
It might also be best to keep the `deviance` since models might cache the value in the struct rather than having to compute it in the fly every time it is called.

Reply:
Good idea. It will automatically handle the weights too.

Reply:
> Not sure for ProbabilityWeights, but should work fine for the other weights supported if not mistaken. The tests should cover the weight cases for sure.

IIRC it should be OK even for probability weights since weights appear both at the numerator and the denominator (therefore multiplying weights doesn't change the result, which is often an issue with probability weights). But I'm afraid we don't have tests for that in GLM (since it only supports frequency weights), maybe you have some in your package?

> Precisely: packages can handle weights through mss(obj) and loglikelihood(obj). There's no need to provide for weights here.

The point is that the new code uses `sum(x -> abs2(x - μ), response(obj))` instead of `deviance`, so it no longer takes weights into account.

Reply:
I am not sure if I ended up supporting `ProbabilityWeights` yet. I know `FrequencyWeights` and `AnalyticalWeights` were supported (basically these worked the same way except in the robust variance covariance estimator which had a subtle difference). If I have time, might help out with developing a simple test suite for the statistical models inheritance. I had a bunch of tests and the computation so one should be able to make a dummy `RegressionModel` for it. In reality one only needs a struct to hold certain values and implement the API for it. No need to actually fit the model so I find the dependency on other packages for fitting it not required.

Reply:
> The point is that the new code uses `sum(x -> abs2(x - μ), response(obj))` instead of `deviance`, so it no longer takes weights into account

I know. I'd removed `deviance` in case a package hadn't defined it. But I reverted this change in the interest of weights in the newest commit.

Reply:
OK. Do you know the formula for weighted Efron R²? I guess we just need to multiply both the numerator and the denominator?

Reply:
> Do you know the formula for weighted Efron R²? I guess we just need to multiply both the numerator and the denominator?

I think so. How are we going to implement it? We can retrieve weights via `weights`, but how do we check if weights were defined in the first place?

Reply:
I think the default is just a one array. Might be nice to follow [FillArrays.jl](https://github.com/JuliaArrays/FillArrays.jl) for it to implement the UnitWeight or something.

Reply:
Let's call `weights` and assume it returns a valid `AbstractVector`. Then it would make sense to define `UnitWeights` in this package, to complement other kinds of weights (https://github.com/JuliaStats/StatsBase.jl/issues/135).

Reply:
Bump. Can you drop the Efron R² for now so that we can merge the PR? We can discuss improvements later.

Reply:
Since https://github.com/JuliaStats/StatsBase.jl/pull/515 we have `UnitWeights`, so we can require models to return this when they are unweighted.